### PR TITLE
Add state sync transaction to debug_trace functions.

### PR DESCRIPTION
**Pull request by 0xsharma**
# Description

Because state sync transaction is added by bor consensus, it will not be included in `debug_traceByBlockNumber` and `debug_tranceByBlockHash`. The problem seems to be that block.Transactions() doesn’t contain state sync transaction.

In this PR, tracing can be possible on state-sync transactions as well. All the debug_trace functions will contain the traces of state-sync transactions, if present.

# Changes

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have added at least 2 reviewer or the whole pos-v1 team
- [x] I have added sufficient documentation in code
- [x] I will be resolving comments - if any - by pushing each fix in a separate commit and linking the commit hash in the comment reply

# Cross repository changes

- [ ] This PR requires changes to heimdall
    - In case link the PR here:
- [ ] This PR requires changes to matic-cli
    - In case link the PR here:

## Testing

- [ ] I have added unit tests
- [ ] I have added tests to CI
- [x] I have tested this code manually on local environment
- [ ] I have tested this code manually on remote devnet using express-cli
- [x] I have tested this code manually on mainnet
- [ ] I have created new e2e tests into express-cli

### CODE REVIEW

**Review by pratikpatil024 — APPROVED** [review]

File: core/vm/interface.go [hunk 1]
```diff
@@ -74,6 +74,8 @@ type StateDB interface {
 	AddPreimage(common.Hash, []byte)
 
 	ForEachStorage(common.Address, func(common.Hash, common.Hash) bool) error
+
+	Finalise(bool)
```
Comment (cffls) [inline core/vm/interface.go:78]: Why is this needed? Changes in this commit only use `Finalise` from statedb right?

File: eth/tracers/api.go [hunk 2]
```diff
@@ -705,7 +795,9 @@ func (api *API) standardTraceBlockToFile(ctx context.Context, block *types.Block
 			canon = false
 		}
 	}
-	for i, tx := range block.Transactions() {
+
+	txs, stateSyncPresent := api.getAllBlockTransactions(ctx, block)
```
Comment (cffls) [inline eth/tracers/api.go:799]: As discussed, we don't want to change the current behavior, which won't include state sync transaction. Could we introduce a new flag to trace API and only include state sync transactions when the flag is set to true?

File: eth/tracers/api_bor.go [hunk 3]
```diff
@@ -115,7 +124,11 @@ func (api *API) traceBorBlock(ctx context.Context, block *types.Block, config *T
 	}
 
 	for indx, tx := range txs {
-		res.Transactions = append(res.Transactions, traceTxn(indx, tx))
+		if stateSyncPresent && indx == len(txs)-1 {
```
Comment (cffls) [inline eth/tracers/api_bor.go:127]: Similar to the comment above, by default, we don't want to include state sync tx even when it presents in the block.

File: eth/tracers/api.go (hunk 2 above)
Comment (0xsharma) [reply eth/tracers/api.go:878]: Done

File: eth/tracers/api_bor.go (hunk 3 above)
Comment (0xsharma) [reply eth/tracers/api_bor.go:127]: Done

File: core/vm/interface.go (hunk 1 above)
Comment (0xsharma) [reply core/vm/interface.go:78]: We need to use Finalise method of the current vmenv state which is possible only if this is enabled. You can try removing this in code and will get the reason why this was added. We can discuss this offline if needed ✅

**Comment by 0xsharma** [comment]
> debug.traceBlockByNumber("0x20FE900",{"borTraceEnabled": true}).length
203
> debug.traceBlockByNumber("0x20FE900",{"borTraceEnabled": false}).length
202
> debug.traceBlockByNumber("0x20FE900").length
202

We can now specify inside the rpc call, if we want the trace of stateSyncTx as well. By default, it is false. This works for all debug trace methods.

**Review by pratikpatil024 — APPROVED** [review]

File: eth/tracers/api.go [hunk 4]
```diff
@@ -911,9 +1126,22 @@ func (api *API) traceTx(ctx context.Context, message core.Message, txctx *Contex
 	// Call Prepare to clear out the statedb access list
 	statedb.Prepare(txctx.TxHash, txctx.TxIndex)
 
-	result, err := core.ApplyMessage(vmenv, message, new(core.GasPool).AddGas(message.Gas()))
-	if err != nil {
-		return nil, fmt.Errorf("tracing failed: %w", err)
+	var result *core.ExecutionResult
+
+	if config.BorTx == nil {
+		config.BorTx = newBoolPtr(false)
```
Comment (cffls) [inline eth/tracers/api.go:1132]: Maybe I am missing something, why do we need `newBoolPtr`?

Comment (0xsharma) [reply eth/tracers/api.go:1132]: Hey, config.BorTx is a *bool and we cannot define it directly AFAIK. So instead of using 3 same lines at multple places. I created a func to give *bool from bool.

Comment (0xsharma) [reply eth/tracers/api.go:1132]: Yes, I too actually was going with `bool`, but then also wanted to keep the same standard as before, therefore used `*bool`.

**Review by cffls — APPROVED** [review]

**Comment by MRabenda** [comment]
@0xsharma @pratikspatil024 should state-sync traces be also available in trace_block?

**Comment by 0xsharma** [comment]
@MRabenda , yes they will be available but as empty traces as state-syncs are system transactions.
